Dockerfile

```diff
@@ -69,76 +39,73 @@ ENV CMS_CFG="$CONFIG_ROOT/cms.yml"
 ENV EDX_PLATFORM_SETTINGS='production'
 
 # Env vars: path
-ENV VIRTUAL_ENV='/edx/app/edxapp/venvs/edxapp'
-ENV PATH="$VIRTUAL_ENV/bin:$PATH"
+ENV VIRTUAL_ENV="/edx/app/edxapp/venvs/edxapp"
+ENV PATH="${VIRTUAL_ENV}/bin:${PATH}"
 ENV PATH="/edx/app/edxapp/edx-platform/node_modules/.bin:${PATH}"
 ENV PATH="/edx/app/edxapp/edx-platform/bin:${PATH}"
 ENV PATH="/edx/app/edxapp/nodeenv/bin:${PATH}"
 
-# Create config directory. Create, define, and switch to working directory.
-RUN mkdir -p "$CONFIG_ROOT"
 WORKDIR /edx/app/edxapp/edx-platform
 
-# Env vars: paver
-# We intentionally don't use paver in this Dockerfile, but Devstack may invoke paver commands
-# during provisioning. Enabling NO_PREREQ_INSTALL tells paver not to re-install Python
-# requirements for every paver command, potentially saving a lot of developer time.
-ENV NO_PREREQ_INSTALL='1'
-
-# Set up a Python virtual environment.
-# It is already 'activated' because $VIRTUAL_ENV/bin was put on $PATH.
-RUN python3.8 -m venv "$VIRTUAL_ENV"
-
-# Install Python requirements.
-# Requires copying over requirements files, but not entire repository.
-COPY requirements requirements
-RUN pip install -r requirements/pip.txt
-RUN pip install -r requirements/edx/base.txt
-
-# Set up a Node environment and install Node requirements.
-# Must be done after Python requirements, since nodeenv is installed
-# via pip.
-# The node environment is already 'activated' because its .../bin was put on $PATH.
-RUN nodeenv /edx/app/edxapp/nodeenv --node=16.14.0 --prebuilt
-RUN npm install -g npm@8.5.x
-COPY package.json package.json
-COPY package-lock.json package-lock.json
-RUN npm set progress=false && npm install
-
-# Copy over remaining parts of repository (including all code).
 COPY . .
 
-# Install Python requirements again in order to capture local projects
-RUN pip install -e .
-
+# Create user before assigning any directory ownership to it.
 RUN useradd -m --shell /bin/false app
 
-USER app
+RUN set -eux; \
```

Comment:
@NIXKnight @adzuci A 40-line run statement? Yikes. This can't be the best way to do this.

Reply:
Sorry, my first comment was too knee-jerk.

I get that you're trying to reduce the final image size, but condensing the majority of the Dockerfile into a single RUN statement defeats the purpose of using Dockerfile at all, which aims to generate cachable images by building them in layers. The reason the old Devstack images are so bad is partly that the entire "Ansible provisioning" step is a single RUN statement, meaning that the images behave horribly with Docker's layer cache. Since 2U is the sole maintainer and user of these images, I'm not super picky about what goes into them, but there's a certain level of best practices that any code in an openedx repo needs to meet.

I recommend reading up on [Dockerfile best practices](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/) to understand why I'm making this comment. Then, in order to optimize the image, I recommend checking out  [multistage builds](https://docs.docker.com/build/building/multi-stage/).